traits/trait_types.py

```diff
@@ -1700,127 +1701,263 @@ def __init__(
+
+    Raises
+    ------
+    TraitError
+        If any of *low*, *high* and *default* is not None, is not a string,
+        not an integer and not a floating-point number.
+
+    """
+    value_types = [
+        value_type
+        for is_numeric, value_type in map(_is_numeric, [low, high])
+        if is_numeric
+    ]
+
+    # For backwards compatibility, assume Float if no value_type is given.
+    if not value_types:
+        warnings.warn(
+            message=(
+                "Unable to infer a value type for this Range instance. "
+                "Using Float(). This warning can be silenced by passing "
+                "value_trait=Int() or value_trait=Float() as appropriate "
+                "at Range creation time. This warning may become an error "
+                "in a future version of Traits. "
+            ),
+            category=DeprecationWarning,
+            stacklevel=3,
+        )
+        trait_type = Float
+    else:
+        trait_type = Float if float in value_types else Int
+
+    return trait_type() if default is None else trait_type(default)
+
 
 class BaseRange(TraitType):
-    """ Defines a trait whose numeric value must be in a specified range.
+    """ A trait whose value must be within a specified range.
     """
 
     def __init__(
         self,
         low=None,
         high=None,
         value=None,
+        low_name=None,
+        high_name=None,
         exclude_low=False,
         exclude_high=False,
+        value_trait=None,
         **metadata
     ):
         """ Creates a Range trait.
 
         Parameters
         ----------
-        low : integer, float or string (i.e. extended trait name)
-            The low end of the range.
-        high : integer, float or string (i.e. extended trait name)
-            The high end of the range.
-        value : integer, float or string (i.e. extended trait name)
-            The default value of the trait.
-        exclude_low : bool
-            Indicates whether the low end of the range is exclusive.
-        exclude_high : bool
-            Indicates whether the high end of the range is exclusive.
-
-        The *low*, *high*, and *value* arguments must be of the same type
-        (integer or float), except in the case where either *low* or *high* is
-        a string (i.e. extended trait name).
-
-        Default Value
-        -------------
-        *value*; if *value* is None or omitted, the default value is *low*,
-        unless *low* is None or omitted, in which case the default value is
-        *high*.
-        """
-        if value is None:
-            if low is not None:
-                value = low
-            else:
-                value = high
-
-        super(BaseRange, self).__init__(value, **metadata)
-
-        vtype = type(high)
-        if (low is not None) and (
-            not issubclass(vtype, (float,) + six.string_types)
-        ):
-            vtype = type(low)
+        low : number or string; optional
```

Comment:
This is misleading - outside of legacy mode it can be anything that matches the value trait.  I'd just use `any` with clarification about legacy behaviour.  Same with `high`.

Reply:
I've reworked the entire docstring, including this. It now says "value or None, optional" for both  `low` and `high`.